### ISSUE TITLE
Ingress docs for Flux webhook receiver are missing important details

### ISSUE THREAD

**Issue by kingdonb** [fluxcd/flux2#2240]
### Describe the bug

I wrote parts of these webhook receiver docs, in "[Expose the webhook receiver](https://fluxcd.io/docs/guides/webhook-receivers/#expose-the-webhook-receiver)" I mentioned cert-manager and proudly stated that you can use the annotations here, but declared it was out of scope to document the whole setup from end to end. I think this use case is common enough (and well-explored) that we should probably document the whole setup from end to end with at least one ingress controller.

And, well, it turns out that the difficulties you encounter when trying to use cert-manager with Flux receivers is interesting enough to be worth covering in docs 👍 

I think that cert-manager is one use case, that may or may not be an issue unique to ingress-nginx, but we could document cert manager, and how to add a network policy permitting an Issuer to manage certificates for Ingress in the flux-system ns. I've been learning Traefik and while I think it might be easier to explain, and can accomplish TLS+LetsEncrypt without cert-manager and without a NetworkPolicy, I don't know if we want to expand this section of the docs any more than necessary.

I know a bit more about this stuff than what I knew when I first wrote this, but I think our community is likely to have better and broader experience than myself, so I started this issue where we can document any other important details that should get a mention in the docs, whether they are for any specific Ingress controller or related to specific CNI implementations that may have quirks one will need to be aware of and work around related to NetworkPolicy resources in Flux. 👍 

### Steps to reproduce

There should be enough detail in the docs to produce a working, secured public Ingress with respect to Flux's NetworkPolicy strategy and guidance around securing the Flux namespace.

### Expected behavior

Let's add a note about how cert-manager changes the network requirements in the namespace as well as explaining how to appropriately permit cert-manager traffic when it's in use for certificate generation and renewal.

### OS / Distro

N/A

### Flux version

v0.24.1

### Flux check

N/A

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by kingdonb** [comment]
Should incorporate information from:

* https://github.com/fluxcd/flux2/issues/1448
* https://github.com/fluxcd/website/issues/359

**Comment by andi0b** [comment]
I was also struggling with it, this seems to work for me, no idea if it's right:

```yaml
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: webhook-receiver
  namespace: flux-system
  annotations:
    cert-manager.io/cluster-issuer: letsencrypt
spec:
  rules:
    - host: fluxwebhook.example
      http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name: webhook-receiver
                port:
                  number: 80
  tls:
    - hosts:
        - fluxwebhook.example
      secretName: webhook-receiver-https
---
kind: NetworkPolicy
apiVersion: networking.k8s.io/v1
metadata:
  name: allow-cert-manager-resolver-reverse
  namespace: cert-manager
spec:
  podSelector:
    matchLabels:
      app.kubernetes.io/instance: cert-manager
  egress:
    - to:
        - namespaceSelector:
            matchLabels:
              acme.cert-manager.io/http01-solver: "true"

---
kind: NetworkPolicy
apiVersion: networking.k8s.io/v1
metadata:
  name: allow-cert-manager-resolver
  namespace: "flux-system"
spec:
  podSelector:
    matchLabels:
      acme.cert-manager.io/http01-solver: "true"
  ingress:
    - from:
        - namespaceSelector:
            matchLabels:              
              app.kubernetes.io/instance: cert-manager
```

**Comment by Klavionik** [comment]
@kingdonb Hi, any plans to update the existing documentation? I've spent a couple of hours today trying to find out why I can't issue a certificate for the webhook receiver ingress. I'm new to all things Kubernetes, and `504 Gateway Time-out` is not a very descriptive error, so it took some time before I reached this issue and realized it's about the network configuration.

For those struggling too, I ended up using this network policy (which simply allows all ingress traffic to http-solver pods in the `flux-system` namespace):
```yaml
---
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-cert-manager-http-solver
  namespace: flux-system
spec:
  policyTypes:
    - Ingress
  podSelector:
    matchLabels:
      acme.cert-manager.io/http01-solver: "true"
  ingress:
    - {}
```

**Comment by andi0b** [comment]
I don’t know how the documentation can be update here (I’m not a contributor), but a simple remark with a link to this issue would probably be enough.

**Comment by fabn** [comment]
+1000 for this, got stuck with 504 for hours as well 🤦

**Comment by Sierra1011** [comment]
Same here, this fixed my unresolved DNS within seconds :heart_eyes:  Thanks @Klavionik !

**Comment by kingdonb** [comment]
The details in this issue, I finally got around to testing today. I was able to make use of all the networkpolicies described here and I think they should go in the docs. I would like to do one more test of cert-manager (to reproduce the original issue) before recommending this change, at least, because I have changed a lot of things, and I honestly think the NetworkPolicy is one of the last things I would have needed to check.

Like for example, the cert-manager challenge is hosted in a parent vcluster, can this policy still be used or does it need some modification then?

I would like to avoid letting perfect get in the way of making these docs actually good. The issue remains open to signify that we are definitely missing a few important use cases. I will try to do some recap and come back to this issue, so we close it.

Thanks for your patience, everyone, but it was harder than I thought to get to the point where the network policy ingress was the only thing preventing me from using a TLS protected webhook on my clusters in the home lab.

I used the suggested most-limited networkpolicy from @andi0b above, that limits traffic to only labeled cert-manager pods.

```
cert-manager   allow-cert-manager-resolver-reverse   app.kubernetes.io/instance=cert-manager   116m
flux-system    allow-cert-manager-resolver           acme.cert-manager.io/http01-solver=true   116m
```

This seems to do the trick, for me! Thanks very much everyone who contributed something to the report here.

**Comment by andi0b** [comment]
@kingdonb the refernced solution is suboptimal as it allows ingress from all pods with the label `acme.cert-manager.io/http01-solver`. Any untrusted pod on an untrusted namespace could set this label and get access. I would suggest adding a namespace to the example.

**Comment by kingdonb** [comment]
@andi0b I was thrown off by this part:

```
egress:
  - to:
      - namespaceSelector:
          matchLabels:
            acme.cert-manager.io/http01-solver: "true"
```

Is that incorrect? Or is the intention that cluster admins would add that label to the namespaces which should be allowed to perform http-01 challenges.

I updated the single rule with a namespace selector to look like this:
```yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-cert-manager-http-solver
  namespace: flux-system
spec:
  policyTypes:
    - Ingress
  podSelector:
    matchLabels:
      acme.cert-manager.io/http01-solver: "true"
  ingress:
  - from:
    - namespaceSelector:
        matchLabels:
          name: cert-manager
```

It still works for me. Is that what you meant? Thanks for the suggestions!

(What is the purpose of the `allow-cert-manager-resolver-reverse` rule? I don't understand and I don't seem to need it)

**Comment by kingdonb** [comment]
Maybe I spoke too soon, the more specific ingress (with `namespaceSelector`) blocked the next cert-manager challenge I made, when I tried to upgrade my cert to the letsencrypt production API - maybe there was a delay propagating the network policy that interfered with my last test.

I'll try adding the reverse policy you suggested and a label where it makes sense, to see if that solves it.

**Comment by kingdonb** [comment]
I haven't been able to make a "good" least-required policy that works - it might be a problem with my cluster's CNI topology:

```
% k -n default get endpoints kubernetes
NAME         ENDPOINTS          AGE
kubernetes   10.96.59.56:6443   14d
yebyen@kingdon-mbp cert-manager % k -n default get svc kubernetes
NAME         TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
kubernetes   ClusterIP   10.95.0.1    <none>        443/TCP   14d
```

The cozystack cluster has a KubeOVN + Cilium network where some services on a parent cluster (this one is KubeVirt/Kamaji) are hosted on the management Kubernetes cluster, so that service doesn't actually live in a namespace on the cluster, it is on the other side of OVN somewhere. I might need to add a more permissive rule.

I think my cluster is not a good example to use to base the receiver docs on, but I guess it is illustrative of the fact that this document can't really cover how to configure Network Policies exhaustively, or even close.

I'll try again on a bare metal cluster or cloud provider hosted (AWS, Azure) and see what else I come up with.

Meanwhile, I think that:
* https://github.com/fluxcd/website/pull/2086
will likely merge without this content, unless I am able to test the solution.

**Comment by kingdonb** [comment]
```yaml
kind: NetworkPolicy
apiVersion: networking.k8s.io/v1
metadata:
  name: allow-cert-manager-resolver-reverse
  namespace: cert-manager
spec:
  podSelector:
    matchLabels:
      app.kubernetes.io/instance: cert-manager
  egress:
    - to:
        - namespaceSelector:
            matchLabels:
              acme.cert-manager.io/http01-solver: "true"
```

I think this should be podSelector, not namespaceSelector - wdyt?

**Comment by kingdonb** [comment]
After further testing I agree with the edit that suggests the `allow-cert-manager-resolver-reverse` policy is unneeded.

On my test cluster, only `allow-cert-manager-resolver` from [this comment](https://github.com/fluxcd/flux2/issues/2240#issuecomment-1321008610) was needed to get Certificates rolling again. Applying only that single policy also did not interfere with the traffic between cert-manager and the Kubernetes API. 🎉 

I think this can be closed. I will have to come back to it, but that's the resolution. I will be sure to include a notice that "this particular network policy may not be appropriate for your environment." I will finish the PR later, but in my case, I guess it was better to not place any restrictions or limits on the traffic allowed in the `cert-manager` namespace. Then the leader election was unblocked.

Sorry for spam, I will try to close this one out soon 🌮
